Sming/system/uart.cpp

```diff
@@ -718,11 +718,61 @@ void uart_set_pins(uart_t* uart, int tx, int rx)
 }
 
 
+#ifdef UART_DEBUG_SNOOP
+
+static void print_stack()
+{
+    register uint32_t sp asm("a1");
+
+    m_printf("Free heap: %u\n", system_get_free_heap_size());
+
+	uint32_t start = sp;
+	uint32_t end = 0x3fffffb0;
+
+    uint32_t pos = 0;
+
+    PSTR_ARRAY(separatorLine, "\n================================================================\n");
+
+    m_puts(separatorLine);
+    for (pos = start; pos < end; pos += 0x10) {
+        uint32_t* values = (uint32_t*)(pos);
+
+        // rough indicator: stack frames usually have SP saved as the second word
+        bool looksLikeStackFrame = (values[2] == pos + 0x10);
+
+        m_printf("%08x:  %08x %08x %08x %08x %c\n",
+            pos, values[0], values[1], values[2], values[3], (looksLikeStackFrame)?'<':' ');
+    }
+    m_puts(separatorLine);
+}
+
+#endif
+
 static void uart_debug_putc(char c)
 {
 	uart_t* uart = uart_get_uart(s_uart_debug_nr);
-	if (uart)
+	if (uart) {
 		uart_write_char(uart, c);
+
+#ifdef UART_DEBUG_SNOOP
+		static char line[16];
+		static unsigned lineLength = 0;
+
+		if(c == '\r' || c == '\n') {
+			if(lineLength >= 2) {
+				if(memcmp(line, "E:M", 2) == 0) {
```

Comment:
@mikee47 Maybe you can change this to "E:M " ( <-- one space after the capital M) just to avoid false positives.